test_conformance/workgroups/test_wg_scan_reduce.cpp

```diff
@@ -0,0 +1,455 @@
+{
+    // Build a kernel string of the form:
+    // __kernel void KERNEL_NAME(global TYPE *input, global TYPE *output) {
+    //     int  tid = get_global_id(0);
+    //     output[tid] = FUNC(input[tid]);
+    // }
+
+    std::ostringstream os;
+    os << "__kernel void " << kernelName << "(global " << type
+       << " *input, global " << type << " *output) {\n";
+    os << "    int tid = get_global_id(0);\n";
+    os << "    output[tid] = " << func << "(input[tid]);\n";
+    os << "}\n";
+    return os.str();
+}
+
+template <typename T> struct TestTypeInfo
+{
+};
+
+template <> struct TestTypeInfo<cl_int>
+{
+    static constexpr const char *deviceName = "int";
+};
+
+template <> struct TestTypeInfo<cl_uint>
+{
+    static constexpr const char *deviceName = "uint";
+};
+
+template <> struct TestTypeInfo<cl_long>
+{
+    static constexpr const char *deviceName = "long";
+};
+
+template <> struct TestTypeInfo<cl_ulong>
+{
+    static constexpr const char *deviceName = "ulong";
+};
+
+template <typename T> struct Add
+{
+    using Type = T;
+    static constexpr const char *opName = "add";
+    static constexpr T identityValue = 0;
+    static T combine(T a, T b) { return a + b; }
+};
+
+template <typename T> struct Max
+{
+    using Type = T;
+    static constexpr const char *opName = "max";
+    static constexpr T identityValue = std::numeric_limits<T>::min();
+    static T combine(T a, T b) { return std::max(a, b); }
+};
+
+template <typename T> struct Min
+{
+    using Type = T;
+    static constexpr const char *opName = "min";
+    static constexpr T identityValue = std::numeric_limits<T>::max();
+    static T combine(T a, T b) { return std::min(a, b); }
+};
+
+template <typename C> struct Reduce
+{
+    using Type = typename C::Type;
+
+    static constexpr const char *testName = "work_group_reduce";
+    static constexpr const char *testOpName = C::opName;
+    static constexpr const char *deviceTypeName =
+        TestTypeInfo<Type>::deviceName;
+    static constexpr const char *kernelName = "test_wg_reduce";
+    static int verify(Type *inptr, Type *outptr, size_t n, size_t wg_size)
+    {
+        for (size_t i = 0; i < n; i += wg_size)
+        {
+            size_t m = n - i;
+            if (m > wg_size) m = wg_size;
+
+            Type result = C::identityValue;
+            for (size_t j = 0; j < m; j++)
+            {
+                result = C::combine(result, inptr[i + j]);
+            }
+
+            for (size_t j = 0; j < m; j++)
+            {
+                if (result != outptr[i + j])
+                {
+                    log_info("%s_%s: Error at %zu\n", testName, testOpName,
+                             i + j);
+                    return -1;
+                }
+            }
+        }
+        return 0;
+    }
+};
+
+template <typename C> struct ScanInclusive
+{
+    using Type = typename C::Type;
+
+    static constexpr const char *testName = "work_group_scan_inclusive";
```

Comment:
It would be nice to have in `testName` the `testOpName` .  Printing output would be more readable than now:
work_group_scan_inclusive_add...
work_group_scan_inclusive int passed
work_group_scan_inclusive uint passed
work_group_scan_inclusive long passed
work_group_scan_inclusive ulong passed
work_group_scan_inclusive_add passed
work_group_scan_inclusive_min...
work_group_scan_inclusive int passed
work_group_scan_inclusive uint passed
work_group_scan_inclusive long passed
work_group_scan_inclusive ulong passed
work_group_scan_inclusive_min passed
work_group_scan_inclusive_max...
work_group_scan_inclusive int passed
work_group_scan_inclusive uint passed
work_group_scan_inclusive long passed
work_group_scan_inclusive ulong passed
work_group_scan_inclusive_max passed
work_group_scan_exclusive_add...
work_group_scan_exclusive int passed
work_group_scan_exclusive uint passed
work_group_scan_exclusive long passed
work_group_scan_exclusive ulong passed
work_group_scan_exclusive_add passed
work_group_scan_exclusive_min...
work_group_scan_exclusive int passed
work_group_scan_exclusive uint passed
work_group_scan_exclusive long passed
work_group_scan_exclusive ulong passed
work_group_scan_exclusive_min passed
work_group_scan_exclusive_max...
work_group_scan_exclusive int passed
work_group_scan_exclusive uint passed
work_group_scan_exclusive long passed
work_group_scan_exclusive ulong passed
work_group_scan_exclusive_max passed

Reply:
Done - nice catch!  I thought I had the output exactly the same but I missed this one.